### PR TITLE
[win32] Use initial cursor size when hiding or showing it

### LINKED ISSUE #2333: [BUG] Console.show_cursor() (and displaying Rich progress bars) clobbers cursor style on Windows
**Describe the bug**

For some context, I originally discovered this with the Rich progress bars in Twine, that whenever one would display, the cursor style would be clobbered to a block cursor, and reported as pypa/twine#902 . However, I subsequently reproduced it in `pip` as and was directed to report it to Rich, and indeed, I can reproduce it with a minimal Rich snippit.

On Windows (tested both 8.1 and 10 on multiple machines) running `rich.console.Console.show_cursor()` clobbers the current cursor style (underbar in the stock Windows terminals, bar in ConEmu) and forces the cursor to display as a block, even if the cursor is already visible, with no way I've found to get it back other than completely closing and restarting the terminal. This also happens whenever I run pip and Twine and Rich progress bars are displayed, unless I specifically disable them through their respective CLIs, which I have to always remember to do, or I have to restart my terminal tab to get my normal cursor back, potentially losing tens of thousands of lines of scrollback and command history.

I've reproduced the same behavior in all environments on my machine I was able to test, as well as the other clean-installed Windows machines I tested (with no obvious common configuration):

* cmd.exe shell running in the default Command Prompt terminal
* Powershell running in the default Powershell terminal
* cmd.exe running in ConEmu
* Git Bash running in ConEmu

(In Git Bash running in its bundled MinTTY, I could not confirm if the behavior can be reproduced since Rich treats it as a not a terminal and the function terminates without doing anything, given `is_terminal` is `False` there and `show_cursor(False)` does nothing.)

Furthermore, I've reproduced the above running the latest version of Rich, 12.4.4, in fresh clean environments with only Rich, pip, Twine and their deps installed, both from PyPI with pip, and from conda-forge with conda.

I've tried numerous methods to restore it once this happens, including modifying my terminal emulator settings, various ANSI escape sequences, the `reset` command in my shell, starting a new `bash` session within an existing one and sshing into another machine, including a Linux box, but nothing works save closing and restarting the terminal.

_However,_ if I `ssh` into a Linux host (prior to triggering the bug) from either ConEmu + Git Bash or MinTTY + Git Bash and run `show_cursor()` there, the cursor both hides and shows correctly, retaining whatever shape I configured in my terminal emulator (and I tested with lines, blocks and underbars). (To note, this works even from MinTTY running Git Bash, despite it not being considered a terminal when running bash on Windows.) The cursor hiding/showing persists (correctly, I assume) even when I exist the remote session, but if I do trigger the bug on Windows, so does the block cursor within and without the ssh session (and `show_cursor()` does not restore the previous one, since the block cursor is set as the current cursor). Therefore, it would seem there is something specific to Windows that is triggering this issue, given I'm running the same shell and same terminal emulator and everything works fine.

Minimal reproducer:

```shell
python -c "import rich.console; rich.console.Console().show_cursor()"
# Cursor now a block
```

**Platform**

What platform (Win/Linux/Mac) are you running on? What terminal software are you using?

### Reproduced issue

All running Python 3.9.13 x64

* {1*) Windows 8.1 Pro x64 + ConEmu 210912 stable + Git Bash 4.4.23(1)-release (x86_64-pc-msys)
* Windows 8.1 Pro x64 + ConEmu 210912 stable + cmd.exe
* Windows 8.1 Pro x64 + Command Prompt + cmd.exe
* Windows 8.1 Pro x64 + Windows Powershell + powershell
* Windows 10 Enterprise x64 + Command Prompt + cmd.exe

<details>

<summary>Windows + ConEmu + Git Bash platform details</summary>

`python -m rich` and `python -m rich.diagnose` display colors and other characters correctly, as far as I can tell.

```console
$ python -m pip list | grep rich
rich               12.4.4

$ python -m rich.diagnose
┌───────────────────────── <class 'rich.console.Console'> ─────────────────────────┐
│ A high level console interface.                                                  │
│                                                                                  │
│ ┌──────────────────────────────────────────────────────────────────────────────┐ │
│ │ <console width=117 ColorSystem.WINDOWS>                                      │ │
│ └──────────────────────────────────────────────────────────────────────────────┘ │
│                                                                                  │
│     color_system = 'windows'                                                     │
│         encoding = 'utf-8'                                                       │
│             file = <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'> │
│           height = 57                                                            │
│    is_alt_screen = False                                                         │
│ is_dumb_terminal = False                                                         │
│   is_interactive = True                                                          │
│       is_jupyter = False                                                         │
│      is_terminal = True                                                          │
│   legacy_windows = True                                                          │
│         no_color = False                                                         │
│          options = ConsoleOptions(                                               │
│                        size=ConsoleDimensions(width=117, height=57),             │
│                        legacy_windows=True,                                      │
│                        min_width=1,                                              │
│                        max_width=117,                                            │
│                        is_terminal=True,                                         │
│                        encoding='utf-8',                                         │
│                        max_height=57,                                            │
│                        justify=None,                                             │
│                        overflow=None,                                            │
│                        no_wrap=False,                                            │
│                        highlight=None,                                           │
│                        markup=None,                                              │
│                        height=None                                               │
│                    )                                                             │
│            quiet = False                                                         │
│           record = False                                                         │
│         safe_box = True                                                          │
│             size = ConsoleDimensions(width=117, height=57)                       │
│        soft_wrap = False                                                         │
│           stderr = False                                                         │
│            style = None                                                          │
│         tab_size = 8                                                             │
│            width = 117                                                           │
└──────────────────────────────────────────────────────────────────────────────────┘
┌─── <class 'rich._windows.WindowsConsoleFeatures'> ────┐
│ Windows features available.                           │
│                                                       │
│ ┌───────────────────────────────────────────────────┐ │
│ │ WindowsConsoleFeatures(vt=False, truecolor=False) │ │
│ └───────────────────────────────────────────────────┘ │
│                                                       │
│ truecolor = False                                     │
│        vt = False                                     │
└───────────────────────────────────────────────────────┘
┌────── Environment Variables ───────┐
│ {                                  │
│     'TERM': 'cygwin',              │
│     'COLORTERM': None,             │
│     'CLICOLOR': None,              │
│     'NO_COLOR': None,              │
│     'TERM_PROGRAM': None,          │
│     'COLUMNS': None,               │
│     'LINES': None,                 │
│     'JPY_PARENT_PID': None,        │
│     'VSCODE_VERBOSE_LOGGING': None │
│ }                                  │
└────────────────────────────────────┘
platform="Windows"
```

</details>

### Could not confirm either way (not considered a terminal)

* (2*) Windows 8.1 Pro x64 + MinTTY 3.1.0 (x86_64-pc-msys) + Git Bash 4.4.23(1)-release (x86_64-pc-msys) + Python 3.9.13

<details>

<summary>Windows + MinTTY + Git Bash platform details</summary>

`python -m rich` and `python -m rich.diagnose` do not display any colors, many of the characters are mojibake/missing, and only work at all without an error when running in UTF-8 mode

```console
$ python -m pip list | grep rich
rich               12.4.4

$ python -m rich.diagnose
+---------------------- <class 'rich.console.Console'> -----------------------+
| A high level console interface.                                             |
|                                                                             |
| +-------------------------------------------------------------------------+ |
| | <console width=79 None>                                                 | |
| +-------------------------------------------------------------------------+ |
|                                                                             |
|     color_system = None                                                     |
|         encoding = 'cp1252'                                                 |
|             file = <_io.TextIOWrapper name='<stdout>' mode='w'              |
|                    encoding='cp1252'>                                       |
|           height = 25                                                       |
|    is_alt_screen = False                                                    |
| is_dumb_terminal = False                                                    |
|   is_interactive = False                                                    |
|       is_jupyter = False                                                    |
|      is_terminal = False                                                    |
|   legacy_windows = True                                                     |
|         no_color = False                                                    |
|          options = ConsoleOptions(                                          |
|                        size=ConsoleDimensions(width=79, height=25),         |
|                        legacy_windows=True,                                 |
|                        min_width=1,                                         |
|                        max_width=79,                                        |
|                        is_terminal=False,                                   |
|                        encoding='cp1252',                                   |
|                        max_height=25,                                       |
|                        justify=None,                                        |
|                        overflow=None,                                       |
|                        no_wrap=False,                                       |
|                        highlight=None,                                      |
|                        markup=None,                                         |
|                        height=None                                          |
|                    )                                                        |
|            quiet = False                                                    |
|           record = False                                                    |
|         safe_box = True                                                     |
|             size = ConsoleDimensions(width=79, height=25)                   |
|        soft_wrap = False                                                    |
|           stderr = False                                                    |
|            style = None                                                     |
|         tab_size = 8                                                        |
|            width = 79                                                       |
+-----------------------------------------------------------------------------+
+--- <class 'rich._windows.WindowsConsoleFeatures'> ----+
| Windows features available.                           |
|                                                       |
| +---------------------------------------------------+ |
| | WindowsConsoleFeatures(vt=False, truecolor=False) | |
| +---------------------------------------------------+ |
|                                                       |
| truecolor = False                                     |
|        vt = False                                     |
+-------------------------------------------------------+
+------ Environment Variables -------+
| {                                  |
|     'TERM': 'xterm',               |
|     'COLORTERM': None,             |
|     'CLICOLOR': None,              |
|     'NO_COLOR': None,              |
|     'TERM_PROGRAM': None,          |
|     'COLUMNS': None,               |
|     'LINES': None,                 |
|     'JPY_PARENT_PID': None,        |
|     'VSCODE_VERBOSE_LOGGING': None |
| }                                  |
+------------------------------------+
platform="Windows"

$ python -X utf8 -m rich.diagnose
┌────────────────────── <class 'rich.console.Console'> ───────────────────────┐
│ A high level console interface.                                             │
│                                                                             │
│ ┌─────────────────────────────────────────────────────────────────────────┐ │
│ │ <console width=79 None>                                                 │ │
│ └─────────────────────────────────────────────────────────────────────────┘ │
│                                                                             │
│     color_system = None                                                     │
│         encoding = 'utf-8'                                                  │
│             file = <_io.TextIOWrapper name='<stdout>' mode='w'              │
│                    encoding='utf-8'>                                        │
│           height = 25                                                       │
│    is_alt_screen = False                                                    │
│ is_dumb_terminal = False                                                    │
│   is_interactive = False                                                    │
│       is_jupyter = False                                                    │
│      is_terminal = False                                                    │
│   legacy_windows = True                                                     │
│         no_color = False                                                    │
│          options = ConsoleOptions(                                          │
│                        size=ConsoleDimensions(width=79, height=25),         │
│                        legacy_windows=True,                                 │
│                        min_width=1,                                         │
│                        max_width=79,                                        │
│                        is_terminal=False,                                   │
│                        encoding='utf-8',                                    │
│                        max_height=25,                                       │
│                        justify=None,                                        │
│                        overflow=None,                                       │
│                        no_wrap=False,                                       │
│                        highlight=None,                                      │
│                        markup=None,                                         │
│                        height=None                                          │
│                    )                                                        │
│            quiet = False                                                    │
│           record = False                                                    │
│         safe_box = True                                                     │
│             size = ConsoleDimensions(width=79, height=25)                   │
│        soft_wrap = False                                                    │
│           stderr = False                                                    │
│            style = None                                                     │
│         tab_size = 8                                                        │
│            width = 79                                                       │
└─────────────────────────────────────────────────────────────────────────────┘
┌─── <class 'rich._windows.WindowsConsoleFeatures'> ────┐
│ Windows features available.                           │
│                                                       │
│ ┌───────────────────────────────────────────────────┐ │
│ │ WindowsConsoleFeatures(vt=False, truecolor=False) │ │
│ └───────────────────────────────────────────────────┘ │
│                                                       │
│ truecolor = False                                     │
│        vt = False                                     │
└───────────────────────────────────────────────────────┘
┌────── Environment Variables ───────┐
│ {                                  │
│     'TERM': 'xterm',               │
│     'COLORTERM': None,             │
│     'CLICOLOR': None,              │
│     'NO_COLOR': None,              │
│     'TERM_PROGRAM': None,          │
│     'COLUMNS': None,               │
│     'LINES': None,                 │
│     'JPY_PARENT_PID': None,        │
│     'VSCODE_VERBOSE_LOGGING': None │
│ }                                  │
└────────────────────────────────────┘
platform="Windows"
```

</details>

### Confirmed working correctly:

*  Ubuntu 20.04.4 LTS via SSH + ConEmu 210912 stable + GNU bash 5.0.17(1)-release (x86_64-pc-linux-gnu) + Python 3.8.10 + ssh from (1*) and (2*)

<details>

<summary>Ubuntu via (1*) platform details</summary>

`python -m rich` and `python -m rich.diagnose` display colors and other characters correctly, as far as I can tell.

```console
$ python -m pip list | grep rich
rich               12.4.4

$ python -m rich.diagnose
╭───────────────────────── <class 'rich.console.Console'> ─────────────────────────╮
│ A high level console interface.                                                  │
│                                                                                  │
│ ╭──────────────────────────────────────────────────────────────────────────────╮ │
│ │ <console width=118 ColorSystem.STANDARD>                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                  │
│     color_system = 'standard'                                                    │
│         encoding = 'utf-8'                                                       │
│             file = <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'> │
│           height = 57                                                            │
│    is_alt_screen = False                                                         │
│ is_dumb_terminal = False                                                         │
│   is_interactive = True                                                          │
│       is_jupyter = False                                                         │
│      is_terminal = True                                                          │
│   legacy_windows = False                                                         │
│         no_color = False                                                         │
│          options = ConsoleOptions(                                               │
│                        size=ConsoleDimensions(width=118, height=57),             │
│                        legacy_windows=False,                                     │
│                        min_width=1,                                              │
│                        max_width=118,                                            │
│                        is_terminal=True,                                         │
│                        encoding='utf-8',                                         │
│                        max_height=57,                                            │
│                        justify=None,                                             │
│                        overflow=None,                                            │
│                        no_wrap=False,                                            │
│                        highlight=None,                                           │
│                        markup=None,                                              │
│                        height=None                                               │
│                    )                                                             │
│            quiet = False                                                         │
│           record = False                                                         │
│         safe_box = True                                                          │
│             size = ConsoleDimensions(width=118, height=57)                       │
│        soft_wrap = False                                                         │
│           stderr = False                                                         │
│            style = None                                                          │
│         tab_size = 8                                                             │
│            width = 118                                                           │
╰──────────────────────────────────────────────────────────────────────────────────╯
╭─── <class 'rich._windows.WindowsConsoleFeatures'> ────╮
│ Windows features available.                           │
│                                                       │
│ ╭───────────────────────────────────────────────────╮ │
│ │ WindowsConsoleFeatures(vt=False, truecolor=False) │ │
│ ╰───────────────────────────────────────────────────╯ │
│                                                       │
│ truecolor = False                                     │
│        vt = False                                     │
╰───────────────────────────────────────────────────────╯
╭────── Environment Variables ───────╮
│ {                                  │
│     'TERM': 'cygwin',              │
│     'COLORTERM': None,             │
│     'CLICOLOR': None,              │
│     'NO_COLOR': None,              │
│     'TERM_PROGRAM': None,          │
│     'COLUMNS': None,               │
│     'LINES': None,                 │
│     'JPY_PARENT_PID': None,        │
│     'VSCODE_VERBOSE_LOGGING': None │
│ }                                  │
╰────────────────────────────────────╯
platform="Linux"
```

</details>

<details>

<summary>Ubuntu via (2*) platform details</summary>

`python -m rich` and `python -m rich.diagnose` display colors and other characters correctly, as far as I can tell.

```console
$ python -m pip list | grep rich
rich               12.4.4

$ python -m rich.diagnose
╭─────────────────────── <class 'rich.console.Console'> ───────────────────────╮
│ A high level console interface.                                              │
│                                                                              │
│ ╭──────────────────────────────────────────────────────────────────────────╮ │
│ │ <console width=80 ColorSystem.STANDARD>                                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│     color_system = 'standard'                                                │
│         encoding = 'utf-8'                                                   │
│             file = <_io.TextIOWrapper name='<stdout>' mode='w'               │
│                    encoding='utf-8'>                                         │
│           height = 24                                                        │
│    is_alt_screen = False                                                     │
│ is_dumb_terminal = False                                                     │
│   is_interactive = True                                                      │
│       is_jupyter = False                                                     │
│      is_terminal = True                                                      │
│   legacy_windows = False                                                     │
│         no_color = False                                                     │
│          options = ConsoleOptions(                                           │
│                        size=ConsoleDimensions(width=80, height=24),          │
│                        legacy_windows=False,                                 │
│                        min_width=1,                                          │
│                        max_width=80,                                         │
│                        is_terminal=True,                                     │
│                        encoding='utf-8',                                     │
│                        max_height=24,                                        │
│                        justify=None,                                         │
│                        overflow=None,                                        │
│                        no_wrap=False,                                        │
│                        highlight=None,                                       │
│                        markup=None,                                          │
│                        height=None                                           │
│                    )                                                         │
│            quiet = False                                                     │
│           record = False                                                     │
│         safe_box = True                                                      │
│             size = ConsoleDimensions(width=80, height=24)                    │
│        soft_wrap = False                                                     │
│           stderr = False                                                     │
│            style = None                                                      │
│         tab_size = 8                                                         │
│            width = 80                                                        │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─── <class 'rich._windows.WindowsConsoleFeatures'> ────╮
│ Windows features available.                           │
│                                                       │
│ ╭───────────────────────────────────────────────────╮ │
│ │ WindowsConsoleFeatures(vt=False, truecolor=False) │ │
│ ╰───────────────────────────────────────────────────╯ │
│                                                       │
│ truecolor = False                                     │
│        vt = False                                     │
╰───────────────────────────────────────────────────────╯
╭────── Environment Variables ───────╮
│ {                                  │
│     'TERM': 'xterm',               │
│     'COLORTERM': None,             │
│     'CLICOLOR': None,              │
│     'NO_COLOR': None,              │
│     'TERM_PROGRAM': None,          │
│     'COLUMNS': None,               │
│     'LINES': None,                 │
│     'JPY_PARENT_PID': None,        │
│     'VSCODE_VERBOSE_LOGGING': None │
│ }                                  │
╰────────────────────────────────────╯
platform="Linux"
```

</details>

### PR DESCRIPTION
**Pull request by DrBenton**
By doing this the users get their original cursor style back when Rich gives control back to the terminal, rather than a "full size" one

Fixes #2333 

## Type of changes

- [x] Bug fix
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [ ] I've added tests for new code. _(new code is tested by existing "show/hide cursor" tests 🙂 )_
- [x] I accept that @willmcgugan may be pedantic in the code review.

#### Before the fix:

We can see the cursor is no longer the same (converted to a "thick" one) after Rich gives control back to the terminal:

https://user-images.githubusercontent.com/722388/173618989-275d0802-fc29-4830-8c19-e63bd3c9b401.mp4

#### After the fix:

Cursor-after-Rich is well and truly the same than cursor-before-Rich:

https://user-images.githubusercontent.com/722388/173619377-9db1605d-a6fb-4df2-aad4-a50b4b873a69.mp4

(I also tested with Powershell, and the fix looks to work there too :-)

### CODE REVIEW

File: tests/test_win32_console.py [hunk 1]
```diff
@@ -37,26 +38,35 @@ def win32_handle():
...
+            cursor_info.bVisible = True
+
+        with mock.patch.object(
+            _win32_console,
+            "GetConsoleScreenBufferInfo",
+            return_value=StubScreenBufferInfo,
+        ) as GetConsoleScreenBufferInfo, mock.patch.object(
+            _win32_console, "GetConsoleCursorInfo", side_effect=stub_console_cursor_info
+        ) as GetConsoleCursorInfo:
+            yield {
```
Comment (DrBenton) [inline tests/test_win32_console.py:54]: the fixture yields the created mocks, which is useful for tests that few need to override the default behaviour :-)

**Review by darrenburns — APPROVED** [review]
Nice work!

File: rich/_win32_console.py [hunk 2]
```diff
@@ -340,6 +364,12 @@ def __init__(self, file: "IO[str]") -> None:
         default_text = GetConsoleScreenBufferInfo(handle).wAttributes
         self._default_text = default_text
 
+        # Store the current cursor size
+        # @link https://docs.microsoft.com/en-us/windows/console/console-cursor-info-str
+        initial_cursor_info = CONSOLE_CURSOR_INFO()
```
Comment (willmcgugan) [inline rich/_win32_console.py:369]: I don't think we should be storing the cursor size in the constructor. Consider if the user changes the cursor size after this object has been created (hardly likely, but possible). Now if they call `hide_cursor` or `show_cursor` the cursor size may be incorrect.

We should add this to `hide_cursor` and `show_cursor`. i.e. get the cursor info first, modify the visible flag, and then set the cursor info in a single operation.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Once change request...

File: rich/_win32_console.py (hunk 2 above)
Comment (DrBenton) [reply rich/_win32_console.py]: Done! :slightly_smiling_face:  [08ea255](https://github.com/Textualize/rich/pull/2339/commits/08ea255fe90e8fe8be069c1d420026cb43030389)

**Pull request merged**